corehq/apps/users/tests/test_role_utils.py

```diff
@@ -118,3 +125,76 @@ def _delete_presets(self):
         for role in UserRole.objects.get_by_domain(self.domain):
             if role.id != self.role1.id:
                 role.delete()
+
+
+class TestCommcareAnalyticsRolesByUser(TestCase):
+
+    DOMAIN = "domain1"
+    USERNAME = "username"
+    PASSWORD = "***"
+
+    @classmethod
+    def setUpClass(cls):
+        cls.domain = create_domain(cls.DOMAIN)
+        cls.user = WebUser.create("domain1", cls.USERNAME, cls.PASSWORD, None, None)
+
+        cls.hq_no_cca_role = cls.create_role("No CCA role", analytics_roles=None)
+
+        cls.limited_cca_roles = [
+            COMMCARE_ANALYTICS_GAMMA,
+            COMMCARE_ANALYTICS_SQL_LAB,
+        ]
+        cls.hq_limited_cca_role = cls.create_role("Limited CCA role", analytics_roles=cls.limited_cca_roles)
+        cls.hq_all_cca_roles_role = cls.create_role("All CCA roles", analytics_roles=[], all_roles=True)
+
+    @classmethod
+    def tearDownClass(cls):
+        for role in UserRole.objects.get_by_domain(cls.domain):
+            role.delete()
+
+        cls.user.delete(deleted_by_domain=cls.domain.name, deleted_by=None)
+        cls.domain.delete()
+
+    def test_admin_user(self):
+        self.user.domain_memberships[0].is_admin = True
+        self.assertTrue(self.user.get_domain_membership(self.DOMAIN).is_admin)
+
+        cca_access = get_commcare_analytics_access_for_user_domain(self.user, self.DOMAIN)
+        self.assertEqual(cca_access['roles'], COMMCARE_ANALYTICS_USER_ROLES)
+        self.assertTrue(cca_access['permissions']['can_edit'])
+        self.assertTrue(cca_access['permissions']['can_view'])
+
+    def test_non_admin_user(self):
+        self.user.set_role(self.domain.name, self.hq_no_cca_role.get_qualified_id())
+        self.assertFalse(self.user.get_domain_membership(self.DOMAIN).is_admin)
+
+        cca_access = get_commcare_analytics_access_for_user_domain(self.user, self.DOMAIN)
+        self.assertEqual(cca_access['roles'], [])
+        self.assertFalse(cca_access['permissions']['can_edit'])
+        self.assertFalse(cca_access['permissions']['can_view'])
+
+    def test_user_has_limited_roles(self):
+        self.user.set_role(self.domain.name, self.hq_limited_cca_role.get_qualified_id())
+        self.assertFalse(self.user.get_domain_membership(self.DOMAIN).is_admin)
+
+        cca_access = get_commcare_analytics_access_for_user_domain(self.user, self.DOMAIN)
+        self.assertEqual(cca_access['roles'], self.limited_cca_roles)
+
+    def test_user_has_all_roles(self):
+        self.user.set_role(self.domain.name, self.hq_all_cca_roles_role.get_qualified_id())
+        self.assertFalse(self.user.get_domain_membership(self.DOMAIN).is_admin)
+
+        cca_access = get_commcare_analytics_access_for_user_domain(self.user, self.DOMAIN)
+        self.assertEqual(cca_access['roles'], COMMCARE_ANALYTICS_USER_ROLES)
+
+    @classmethod
+    def create_role(cls, role_name, analytics_roles=None, all_roles=False):
```

Comment:
nit: this method would be better placed after the setUpClass method since its used there.

Additionally, it could be a private method.

Reply:
Making it private, although I like to have private methods at the bottom of the class. It's also the pattern in the file.

Reply:
Oh, I don't know how private functions should be aligned considering the newspaper metaphor (from clean code book) but if that is recommended then all good.

Reply:
Grammar nit: In Python, methods and attributes that are prefixed with a single underscore are protected (and they are protected by convention, not by syntax). Methods and variables that are prefixed with a double underscore are private (and they are private by syntax, in that you can't access them using `my_instance.__my_private_variable` ... although it is still possible to get to them using `my_instance._MyClass__my_private_variable` (because Guido van Rossum treats developers as adults, and allows them to break conventions if they believe it is important, although it will require writing super ugly code to do it :) ).

See [diveintopython.org](https://diveintopython.org/learn/classes/variables#public-protected-and-private-attributes-and-properties-in-python) for runnable examples, and [betterprogramming.pub](https://betterprogramming.pub/public-private-and-protected-access-modifiers-in-python-9024f4c1dd4) for an example of accessing private methods and attributes.